benchmark/httpapi.py

```diff
@@ -0,0 +1,215 @@
+
+from klein import Klein
+
+from zope.interface import implementer
+
+from ._interfaces import IBackend
+
+
+class ResultNotFound(Exception):
+    """
+    Exception indicating that a result with a given identifier is not found.
+    """
+
+
+@implementer(IBackend)
+class DummyBackend(object):
+    """
+    The backend that simply drops all results.
+    """
+    def store(self, result):
+        """
+        Pretend to store a single benchmarking result.
+
+        :param dict result: The result in the JSON compatible format.
+        :return: A Deferred that produces an identifier for the stored
+            result.
+        """
+        return succeed(uuid4().hex)
+
+    def retrieve(self, id):
+        """
+        Just fail.
+        """
+        return fail(ResultNotFound())
+
+    def query(self, filter):
+        """
+        Return an empty result.
+        """
+        return succeed(list())
+
+    def delete(self, id):
+        """
+        Just fail.
+        """
+        return fail(ResultNotFound())
+
+
+@implementer(IBackend)
+class InMemoryBackend(object):
+    """
+    The backend that simply drops all results.
+
+    :ivar dict results: Stored results by their identifiers.
+    """
+    def __init__(self):
+        self.results = {}
+
+    def store(self, result):
+        """
+        Store a single benchmarking result and return its identifier.
+
+        :param dict result: The result in the JSON compatible format.
+        :return: A Deferred that produces an identifier for the stored
+            result.
+        """
+        id = uuid4().hex
+        self.results[id] = result
+        return succeed(id)
+
+    def retrieve(self, id):
+        """
+        Retrive a result by the given identifier.
+        """
+        try:
+            return self.results[id]
+        except KeyError:
+            return fail(ResultNotFound())
+
+    def query(self, filter):
+        """
+        Return matching results.
+        """
+        matching = [
+            r for r in self.results.viewvalues()
+            if filter.viewitems() <= r.viewitems()
+        ]
+        return succeed(matching)
+
+    def delete(self, id):
+        """
+        Delete a result by the given identifier.
+        """
+        try:
+            del self.results[id]
+        except KeyError:
+            return fail(ResultNotFound())
+
+
+class BenchmarkAPI_V1(object):
+    """
+    API for storing and accessing benchmarking results.
+
+    :ivar IBackend backend: The backend for storing the results.
+    """
+    app = Klein()
+    version = 1
+
+    def __init__(self, backend):
+        """
+        :param IBackend backend: The backend for storing the results.
+        """
+        self.backend = backend
+
+    @app.handle_errors(ResultNotFound)
+    def _not_found(self, request, failure):
+        request.setResponseCode(404)
+        return ""
+
+    @app.route("/submit", methods=['POST'])
+    def submit(self, request):
```

Comment:
Seems like your comment somehow got posted incomplete.

I had many doubts over the query interface.
In the end I went for POST, because I wanted to submit JSON as a filter / template.
Doing that through arguments in a GET request is not as natural as just placing JSON into a body of the POST request, despite the GET being more natural for a query in general.

I did some searching on stackoverflow and the "consensus" there seems to be that using the POST for a query is not unacceptable...